pallets/subtensor/src/coinbase/tao_weight.rs

```diff
@@ -0,0 +1,180 @@
+        let tick =
+            U96F32::saturating_from_num(1u64).safe_div(U96F32::saturating_from_num(3_000_000u64));
+        let tao_weight = Self::get_tao_weight();
+
+        let min_weight_u64 = Self::convert_float_to_u64(U96F32::saturating_from_num(0.09));
+        let max_weight_u64 = Self::convert_float_to_u64(U96F32::saturating_from_num(0.18));
+
+        let mut next_tao_weight_u64 = TaoWeight::<T>::get();
+
+        let tao_scale = U96F32::saturating_from_num(1_000_000_000u64);
+
+        // 3)  Compare difference between current and expected reserves.
+        // 3a) We’re under‐filled (need to sell less / weight ↓).
+        if current_reserves < expected_reserves {
+            let diff_raw = expected_reserves.saturating_sub(current_reserves);
+            // Convert to actual TAO units by dividing by 1e9
+            let diff_tao = diff_raw.safe_div(tao_scale);
+            // Apply 1 tick per TAO difference
+            let adjustment = diff_tao.saturating_mul(tick);
+            let next_tao_weight_float = tao_weight.saturating_sub(adjustment);
+
+            next_tao_weight_u64 = Self::convert_float_to_u64(next_tao_weight_float);
+
+            log::debug!(
+                "TAO reserves under-filled: current_reserves: {}, expected_reserves: {}, diff_tao: {}, adjustment: {}, next_tao_weight_float: {}, next_tao_weight_u64: {}",
+                current_reserves,
+                expected_reserves,
+                diff_tao,
+                adjustment,
+                next_tao_weight_float,
+                next_tao_weight_u64
+            );
+        // 3b) We’re over‐filled (need to sell more / weight ↑).
+        } else if expected_reserves < current_reserves {
+            let diff_raw = current_reserves.saturating_sub(expected_reserves);
+            // Convert to actual TAO units by dividing by 1e9
+            let diff_tao = diff_raw.safe_div(tao_scale);
+            // Apply 1 tick per TAO difference
+            let adjustment = diff_tao.saturating_mul(tick);
+            let next_tao_weight_float = tao_weight.saturating_add(adjustment);
+
+            next_tao_weight_u64 = Self::convert_float_to_u64(next_tao_weight_float);
+
+            log::debug!(
+                "TAO reserves over-filled: current_reserves: {}, expected_reserves: {}, diff_tao: {}, adjustment: {}, next_tao_weight_float: {}, next_tao_weight_u64: {}",
+                current_reserves,
+                expected_reserves,
+                diff_tao,
+                adjustment,
+                next_tao_weight_float,
+                next_tao_weight_u64
+            );
+        }
+
+        // 4) Clamp TAO weight and update storage.
+        if next_tao_weight_u64 > max_weight_u64 {
+            TaoWeight::<T>::set(max_weight_u64);
+        } else if next_tao_weight_u64 < min_weight_u64 {
+            TaoWeight::<T>::set(min_weight_u64);
+        } else {
+            TaoWeight::<T>::set(next_tao_weight_u64);
+        }
```

Comment:
why not use .min and .max?

Reply:
also we have a `clamp_value`, we can make it more generic and reuse here so it's more readable